peer.go

```diff
@@ -159,11 +161,50 @@ func (l *PeerList) Copy() map[string]*Peer {
 
 	listCopy := make(map[string]*Peer)
 	for k, v := range l.peersByHostPort {
-		listCopy[k] = v
+		listCopy[k] = v.Peer
 	}
 	return listCopy
 }
 
+// exists checks if a hostport exists in the peer list.
+func (l *PeerList) exists(hostPort string) (*peerScore, bool) {
+	l.RLock()
+	ps, ok := l.peersByHostPort[hostPort]
+	l.RUnlock()
+
+	return ps, ok
+}
+
+// UpdatePeerHeap updates the peer heap.
+func (l *PeerList) UpdatePeerHeap(p *Peer) {
+	if ps, ok := l.exists(p.hostPort); ok {
+		l.Lock()
```

Comment:
don't need to lock while scoring. You need a read lock on `ps` when reading the score but don't need to lock the whole peer list.